src/wms/src/main/java/org/geoserver/wms/WMS.java

```diff
@@ -1756,14 +1877,95 @@ public Filter getTimeElevationToFilter(
      * @return builded filter
      */
+    @Deprecated
     public Filter getDimensionsToFilter(
             final Map<String, String> rawKVP, final FeatureTypeInfo typeInfo) {
+        DimensionFilterBuilder builder = new DimensionFilterBuilder(ff);
+        return getCustomDimensionFilter(rawKVP, typeInfo, builder);
+    }
+
+    /** Builds filters for the custom dimensions */
+    public Filter getCustomDimensionFilter(
+            Map<String, String> rawKVP, FeatureTypeInfo typeInfo, DimensionFilterBuilder builder) {
+        CustomDimensionFilterConverter converter = getCustomDimensionFilterConverter(typeInfo);
+        return converter.getDimensionsToFilter(rawKVP, builder);
+    }
+
+    private CustomDimensionFilterConverter getCustomDimensionFilterConverter(
+            FeatureTypeInfo typeInfo) {
         CustomDimensionFilterConverter.DefaultValueStrategyFactory defaultValueStrategyFactory =
                 (resource, dimensionName, dimensionInfo) ->
                         getDefaultValueStrategy(resource, dimensionName, dimensionInfo);
         final CustomDimensionFilterConverter converter =
-                new CustomDimensionFilterConverter(defaultValueStrategyFactory, ff);
-        return converter.getDimensionsToFilter(rawKVP, typeInfo);
+                new CustomDimensionFilterConverter(typeInfo, defaultValueStrategyFactory);
+        return converter;
+    }
+
+    /**
+     * Validates that the provided custom dimension values are within the allowed range, eventually
+     * throws ServiceException#INVALID_DIMENSION_VALUE if not.
+     *
+     * @param times
+     * @param elevations
+     * @param mapLayerInfo
+     * @param request
+     */
+    public void validateRasterDimensions(
+            List<Object> times,
+            List<Object> elevations,
+            MapLayerInfo mapLayerInfo,
+            GetMapRequest request)
+            throws IOException {
+        // value validation and exception throwing enabled?
+        if (!exceptionOnInvalidDimension()) return;
+
+        CoverageInfo coverage = mapLayerInfo.getCoverage();
+        Map<String, DimensionInfo> dimensions =
+                coverage.getMetadata().entrySet().stream()
+                        .filter(e -> e.getValue() instanceof DimensionInfo)
+                        .filter(e -> ((DimensionInfo) e.getValue()).isEnabled())
+                        .filter(e -> !((DimensionInfo) e.getValue()).isNearestMatchEnabled())
+                        .collect(
+                                Collectors.toMap(
+                                        e -> e.getKey(), e -> (DimensionInfo) e.getValue()));
+        ReaderDimensionsAccessor accessor =
+                new ReaderDimensionsAccessor(
+                        (GridCoverage2DReader) mapLayerInfo.getCoverageReader());
+        for (Map.Entry<String, DimensionInfo> entry : dimensions.entrySet()) {
+            String key = entry.getKey().replaceFirst("dim_", "");
```

Comment:
dim_ magic string should be defined statically somewhere

Reply:
It is, but in a very painful way: ResourceInfo.VECTOR_CUSTOM_DIMENSION_PREFIX. Will try to collect it as a local constant instead (besides, to make things even messier, "dim_" has a different meaning in GetMap, used only to prefix custom dimensions, while in the configuration we use two different prefixes for vector and raster custom dimensions... oh my).